hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java

```diff
@@ -0,0 +1,249 @@
+ *
+ * <p>However, storing them in a queue of this nature does not provide efficient access to the data itself. For this
+ * reason, in-memory data structures are used to provide efficient access to the data. These data structures are rebuilt
+ * after reconnect or restart, and kept in sync with the data in state.
+ *
+ * <p>Mutation methods must be called during startup, reconnect, or on the "handle" thread. Getters may be called from
+ * any thread.
+ */
+@Singleton
+public class RecordCacheImpl implements HederaRecordCache {
+    /**
+     * An item stored in the {@link #histories} cache. These histories are stored in a map keyed by transaction ID.
+     *
+     * @param nodeIds The IDs of every node that submitted a transaction with the txId that came to consensus and was
+     *                handled.
+     * @param records Every {@link TransactionRecord} handled for every transaction that came to consensus with the txId
+     */
+    private record History(@NonNull Set<Long> nodeIds, @NonNull List<TransactionRecord> records) {
+
+        History() {
+            this(new HashSet<>(), new ArrayList<>());
+        }
+    }
+
+    /** Gives access to the current working state. */
+    private final WorkingStateAccessor workingStateAccessor;
+    /** Used for looking up the max transaction duration window. To be replaced by some new config object */
+    private final GlobalDynamicProperties props;
+    /** Used for answering queries about receipts to include those that have not been handled but are known */
+    private final DeduplicationCache deduplicationCache;
+    /**
+     * A map of transaction IDs to the histories of all transactions that came to consensus with that ID. This data
+     * structure is rebuilt during reconnect or restart.
+     */
+    private final Map<TransactionID, History> histories;
+    /**
+     * A secondary index that maps from an AccountID of the payer account to a set of transaction IDs that were
+     * submitted by this payer. This is only needed for answering such queries. Ideally such queries would exist on the
+     * mirror node instead.
+     */
+    private final Map<AccountID, Set<TransactionID>> payerToTransactionIndex = new ConcurrentHashMap<>();
+
+    @Inject
+    public RecordCacheImpl(
+            @NonNull final DeduplicationCache deduplicationCache,
+            @NonNull final WorkingStateAccessor workingStateAccessor,
+            @NonNull final GlobalDynamicProperties props) {
+        this.deduplicationCache = requireNonNull(deduplicationCache);
+        this.workingStateAccessor = requireNonNull(workingStateAccessor);
+        this.props = requireNonNull(props);
+        this.histories = new ConcurrentHashMap<>();
+
+        rebuild();
+    }
+
+    /**
+     * Rebuild the internal data structures based on the current working state. Called during startup and during
+     * reconnect.
+     */
+    public void rebuild() {
+        final var queue = getQueue();
```

Comment:
Desirable to call `histories.clear()` here?

Reply:
Yes! Good catch, I will fix and add the unit test to prove it.